frontend/src/pages/Compare.tsx

```diff
@@ -47,23 +58,68 @@ export default function Compare(props: CompareProps) {
           mode: 'error',
         });
       },
-      onSuccess: () => props.updateBanner({}),
+      onSuccess: data => {
+        // Set the version based on the runs included.
+        let version: CompareVersion = CompareVersion.InvalidRunCount;
+        if (data && data.length >= 2 && data.length <= 10) {
+          for (const run of data) {
+            const runVersion = run.run?.pipeline_spec?.hasOwnProperty('pipeline_manifest')
+              ? CompareVersion.V2
+              : CompareVersion.V1;
+            if (version === CompareVersion.InvalidRunCount) {
+              version = runVersion;
+            } else if (version !== runVersion) {
+              version = CompareVersion.Mixed;
+            }
+          }
+        }
+
+        // Update banner based on feature flag, run versions, and run count.
+        if (isFeatureEnabled(FeatureKey.V2_ALPHA) && version === CompareVersion.InvalidRunCount) {
+          props.updateBanner({
+            additionalInfo:
+              'At least two runs and at most ten runs must be selected to view the Run Comparison page.',
+            message:
+              'Error: failed loading the Run Comparison page. Click Details for more information.',
+            mode: 'error',
+          });
+        } else if (isFeatureEnabled(FeatureKey.V2_ALPHA) && version === CompareVersion.Mixed) {
+          props.updateBanner({
+            additionalInfo:
+              'The selected runs are a mix of V1 and V2.' +
+              ' Please select all V1 or all V2 runs to view the associated Run Comparison page.',
+            message:
+              'Error: failed loading the Run Comparison page. Click Details for more information.',
+            mode: 'error',
+          });
+        } else if (!isFeatureEnabled(FeatureKey.V2_ALPHA) && version === CompareVersion.V2) {
+          props.updateBanner({
+            additionalInfo:
+              'The selected runs are all V2, but the V2_ALPHA feature flag is disabled.' +
+              ' The V1 page will not show any useful information for these runs.',
+            message:
+              'Info: enable the V2_ALPHA feature flag in order to view the updated Run Comparison page.',
+            mode: 'info',
+          });
+        } else {
+          props.updateBanner({});
+        }
```

Comment:
I tend to put logics like this outside of `onSuccess`, `onError` callback. The main reason is that `updateBanner` scattered across multiple places, while we want them to be together for easy management. I suggest using `useEffect`:

```
useEffect(() => {
    if (error) {
          // update banner with error message
    } else if (version === CompareVersion.InvalidRunCount) {
          // update banner with error message
    } else if (version === CompareVersion.Mixed) {
          // update banner with error message
    } else if (version === CompareVersion.V2 && !isFeatureEnabled(FeatureKey.V2_ALPHA)) {
          // update banner with error message
    } else {
          props.updateBanner({});
    }
 
}, [compareVersion, error])
```

More information: https://reactjs.org/docs/hooks-effect.html

In this case, you don't need to worry about child page `CompareV1` wiping banner because we don't show V1 or V2 page when we have an error banner in Compare page.

Reply:
You're right, this is a better solution! I mistakenly ruled this out earlier, but it will only run when the dependencies change, so it won't cause an infinite refresh unlike calling `props.updateBanner` directly in the function body.

I updated the code to incorporate `useEffect` in this commit: ___.